scanspec/service.py

```diff
@@ -1,19 +1,114 @@
+import graphql
 from aiohttp import web
-from apischema.graphql import graphql_schema
+from apischema.graphql import graphql_schema, resolver
 from graphql_server.aiohttp.graphqlview import GraphQLView, _asyncify
+from numpy import array2string, dtype, float64, frombuffer, ndarray
 
+from scanspec.core import Path
 from scanspec.specs import Spec
 
+# See confluence page for clarity on naming conventions and data structure:
+# https://confluence.diamond.ac.uk/display/SSCC/Data+Structure+and+Naming+Conventions
 
+
+# The lowest query level
+@dataclass
+class Points:
+    def __init__(self, points: Optional[ndarray]):
+        self._points = points
+
+    @resolver
+    def string(self) -> Optional[str]:
+        return array2string(self._points)
+
+    @resolver
+    def float_list(self) -> Optional[List[float]]:
+        if self._points is None:
+            return None
+        else:
+            return self._points.tolist()
+
+    @resolver
+    def b64(self) -> Optional[str]:
+        if self._points is None:
+            return None
+        else:
+            # make sure the data is sent as float64
+            assert dtype(self._points[0]) == dtype(float64)
+            return base64.b64encode(self._points.tobytes()).decode("utf-8")
+
+    # Self b64 decoder for testing purposes
+    @resolver
+    def b64Decode(self) -> Optional[str]:
+        if self._points is None:
+            return None
+        else:
+            r = dtype(self._points[0])
+            s = base64.decodebytes(base64.b64encode(self._points.tobytes()))
+            t = frombuffer(s, dtype=r)
+            return array2string(t)
+
+
+@dataclass
+class AxisFrames:
+    axis: str
+    lower: Optional[Points]
+    midpoints: Optional[Points]
+    upper: Optional[Points]
+
+
+# The highest query level
+@dataclass
+class PointsRequest:
+    axes: List[AxisFrames]
+    num_points: int
+
+
+# Chacks that the spec will produce a valid scan
 def validate_spec(spec: Spec) -> Any:
     # apischema will do all the validation for us
     return spec.serialize()
 
 
-schema = graphql_schema(query=[validate_spec])
+# Returns a full list of points for each axis in the scan
+# TODO adjust to return a reduced set of scanPoints
+def get_points(spec: Spec) -> PointsRequest:
+
+    dims = spec.create_dimensions()  # Grab dimensions from spec
+    path = Path(dims)  # Convert to a path
+    num_points = len(path)  # Capture the length of the path
+
+    # WARNING: path object is consumed after this line
+    chunk = path.consume()
+
+    # POINTS #
+    scan_points = []
+    # For every dimension of the scan...
+    for axis in chunk.midpoints:
+        # Extract the midpoints and the upper and lower bounds
+        a = AxisFrames(
+            axis,
+            Points(chunk.lower.get(axis)),
+            Points(chunk.midpoints.get(axis)),
+            Points(chunk.upper.get(axis)),
+        )
+        # Append the information as a list of frames per axis
+        scan_points.append(a)
```

Comment:
I think with nice black formatting list comprehension would be more readable here. Also suggest getting axes from `specs.axes()` because it seems more obvious that you are getting the axes than `axis in chunk.midpoints`
```suggestion
    scan_points = [
        AxisFrames(
            axis,
            Points(chunk.lower.get(axis)),
            Points(chunk.midpoints.get(axis)),
            Points(chunk.upper.get(axis)),
        )
        for axis in spec.axes()
    ]
```

Reply:
Its a bit harder to understand for someone who's not very familiar with python, but I agree it looks much tidier. 

I have updated this in the next revision